### ISSUE TITLE
How to add a new activation function to YOLOv8 that is not already included in PyTorch？

### ISSUE THREAD

**Issue by binwellwellwell** [ultralytics/ultralytics#13137]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

How to add a new activation function to YOLOv8 that is not already included in PyTorch？

### Additional

no

**Comment by glenn-jocher** [comment]
Hi there! To add a new activation function to YOLOv8, you can define your custom activation function in Python and then integrate it into the YOLOv8 model. Here's a quick example:

1. Define your custom activation function:
    ```python
    import torch
    import torch.nn as nn

    class CustomActivation(nn.Module):
        def forward(self, x):
            return x * torch.sigmoid(x)  # Example: Swish activation
    ```

2. Integrate it into the YOLOv8 model:
    ```python
    from ultralytics import YOLO

    model = YOLO('yolov8n.yaml')  # Load the model configuration
    model.model[-1].activation = CustomActivation()  # Replace the activation function
    ```

Now, your YOLOv8 model will use the custom activation function. Happy coding! 🚀

**Comment by SutirthaChakraborty** [comment]
Can this be done on the ultralytics website training settings?

here?
<img width="800" alt="image" src="https://github.com/ultralytics/ultralytics/assets/47362033/d9bb1e37-ab7d-4fc7-be49-babde342fb9f">

**Comment by glenn-jocher** [comment]
@SutirthaChakraborty hello! Thank you for reaching out with your questions. Let's address them one by one:

1. **Custom Activation Function on Ultralytics Website**:
   Currently, the Ultralytics website training settings do not support adding custom activation functions directly through the interface. Custom modifications like these need to be done in the codebase. If you need to implement a custom activation function, you'll have to modify the model architecture in your local environment.

2. **Sigmoid Activation Function for Independent Class Predictions**:
   Yes, you can use a sigmoid activation function to treat each class prediction independently. This is typically used in multi-label classification tasks where each class prediction is independent of the others. Here's an example of how you might define such an activation function:
   ```python
   import torch.nn as nn

   class SigmoidActivation(nn.Module):
       def forward(self, x):
           return torch.sigmoid(x)
   ```

3. **Error: `'DetectionModel' object is not subscriptable`**:
   The error you're encountering indicates that the model object does not support indexing. To modify the activation function, you need to correctly access the specific layer in the model. Here's an updated example:
   ```python
   from ultralytics import YOLO
   import torch.nn as nn

   class CustomActivation(nn.Module):
       def forward(self, x):
           return x * torch.sigmoid(x)  # Example: Swish activation

   # Load the model configuration
   model = YOLO('yolov8n.yaml')

   # Assuming the activation function is part of the last layer
   if hasattr(model.model[-1], 'activation'):
       model.model[-1].activation = CustomActivation()
   else:
       print("The specified layer does not have an 'activation' attribute.")
   ```

Please ensure you are using the most recent versions of `torch` and `ultralytics` packages. You can upgrade them using:
```bash
pip install --upgrade torch ultralytics
```

If you continue to experience issues, please provide a minimum reproducible code example so we can better assist you. You can find more details on creating a reproducible example here: [Minimum Reproducible Example](https://docs.ultralytics.com/help/minimum_reproducible_example).

Feel free to reach out if you have any more questions! 😊
